universal-application-tool-0.0.1/app/modules/SecurityModule.java

```diff
@@ -103,20 +113,79 @@ protected ProfileFactory provideProfileFactory(
     return new ProfileFactory(clock, dbContext, httpContext);
   }
 
+  @Provides
+  @Singleton
   protected FakeAdminClient fakeAdminClient(ProfileFactory profileFactory) {
     return new FakeAdminClient(profileFactory);
   }
 
+  @Named("idcs")
+  @Provides
+  @Nullable
+  @Singleton
+  protected OidcClient provideIDCSClient(ProfileFactory profileFactory) {
+    if (!this.configuration.hasPath("idcs.client_id") || !this.configuration.hasPath("idcs.secret")) {
+      return null;
+    }
+    OidcConfiguration config = new OidcConfiguration();
+    config.setClientId(this.configuration.getString("idcs.client_id"));
+    config.setSecret(this.configuration.getString("idcs.secret"));
+    config.setDiscoveryURI(this.configuration.getString("idcs.discovery_uri"));
+    config.setResponseMode("form_post");
+    config.setResponseType("id_token");
+    config.setUseNonce(true);
+    config.setWithState(false);
+    OidcClient client = new OidcClient(config);
+    client.setCallbackUrl(baseUrl + "/callback");
```

Comment:
Question (blocking): can this use the reverse router instead of string concat?

Reply:
Not really - the client needs to add the magic string "OidcClient" (or "AdClient"), and the reverse router doesn't have a way to say "give me this without the suffix".

Reply:
Bummer